src/ResourceManager/Compute/Commands.Compute/Strategies/AsyncCmdletExtensions.cs

```diff
@@ -79,20 +81,39 @@ public static void WriteVerbose(this IAsyncCmdlet cmdlet, string message, params
                 });
         }
 
-        sealed class AsyncCmdlet : IAsyncCmdlet
+        sealed class AsyncCmdlet<T> : IAsyncCmdlet
+            where T : PSCmdlet
         {
             public SyncTaskScheduler Scheduler { get; } = new SyncTaskScheduler();
 
-            readonly Cmdlet _Cmdlet;
+            readonly T _Cmdlet;
 
             public List<ITaskProgress> TaskProgressList { get; } 
                 = new List<ITaskProgress>();
 
-            public AsyncCmdlet(Cmdlet cmdlet)
+            public AsyncCmdlet(T cmdlet)
             {
                 _Cmdlet = cmdlet;
             }
 
+            public bool WhatIf
+                => _Cmdlet.MyInvocation.BoundParameters.ContainsKey("WhatIf");
+
+            public IEnumerable<KeyValuePair<string, object>> Parameters
+            {
+                get
+                {
+                    var psName = _Cmdlet.ParameterSetName;
+                    return typeof(T)
+                        .GetProperties()
+                        .Where(p => p
+                            .GetCustomAttributes(false)
```

Comment:
we should check for null here, what if there is a random class property that does not have any custom attributes... will the code here throw a null ptr exception?

Reply:
`Type.GetProperties()` and `PropertyInfo.GetCustomAttributes()` never returns `null`. 
- https://msdn.microsoft.com/en-us/library/kyaxdd3x(v=vs.110).aspx
- https://msdn.microsoft.com/en-us/library/kff8s254(v=vs.110).aspx